algorithms/kernel/kernel_function/kernel_function_rbf_dense_default_impl.i

```diff
@@ -133,190 +166,105 @@ services::Status KernelImplRBF<defaultDense, algorithmFPType, cpu>::computeInter
             const algorithmFPType * dataA2 = const_cast<algorithmFPType *>(mtA2.get());
 
-            for (size_t i = startRow, ii = 0; i < finishRow; ++i, ++ii)
+            for (size_t i = 0; i < nRowsInBlock2; ++i)
             {
                 sqrDataA2[i] = zero;
                 for (size_t j = 0; j < nFeatures; ++j)
                 {
-                    sqrDataA2[i] += dataA2[ii * nFeatures + j] * dataA2[ii * nFeatures + j];
+                    sqrDataA2[i] += dataA2[i * nFeatures + j] * dataA2[i * nFeatures + j];
                 }
             }
-        });
-
-        daal::threader_for(blockNum1, blockNum1, [&](const size_t iBlock1) {
-            DAAL_INT numRowsInBlock1 = (iBlock1 != blockNum1 - 1) ? blockSize : nVectors1 - iBlock1 * blockSize;
-            DAAL_INT startRow1       = iBlock1 * blockSize;
-            DAAL_INT finishRow1      = startRow1 + numRowsInBlock1;
-
-            ReadRows<algorithmFPType, cpu> mtA1(*const_cast<NumericTable *>(a1), startRow1, numRowsInBlock1);
-            DAAL_CHECK_BLOCK_STATUS_THR(mtA1);
-            const algorithmFPType * dataA1 = const_cast<algorithmFPType *>(mtA1.get());
-
-            WriteRows<algorithmFPType, cpu> mtR(r, startRow1, numRowsInBlock1);
-            DAAL_CHECK_BLOCK_STATUS_THR(mtR);
-            algorithmFPType * dataR = mtR.get();
-
-            daal::threader_for(blockNum2, blockNum2, [&, nVectors2, blockNum2](const size_t iBlock2) {
-                DAAL_INT numRowsInBlock2 = (iBlock2 != blockNum2 - 1) ? blockSize : nVectors2 - iBlock2 * blockSize;
-                DAAL_INT startRow2       = iBlock2 * blockSize;
-
-                ReadRows<algorithmFPType, cpu> mtA2(*const_cast<NumericTable *>(a2), startRow2, numRowsInBlock2);
-                DAAL_CHECK_BLOCK_STATUS_THR(mtA2);
-                const algorithmFPType * dataA2 = const_cast<algorithmFPType *>(mtA2.get());
-
-                Blas<algorithmFPType, cpu>::xxgemm(&trans, &notrans, &numRowsInBlock2, &numRowsInBlock1, (DAAL_INT *)&nFeatures, &negTwo, dataA2,
-                                                   (DAAL_INT *)&nFeatures, dataA1, (DAAL_INT *)&nFeatures, &zero, dataR + iBlock2 * blockSize,
-                                                   (DAAL_INT *)&nVectors2);
-            });
 
-            for (size_t i = startRow1, ii = 0; i < finishRow1; ++i, ++ii)
+            DAAL_INT lda = nFeatures;
+            DAAL_INT ldb = nFeatures;
+            DAAL_INT ldc = blockSize;
+            Blas<algorithmFPType, cpu>::xxgemm(&trans, &notrans, &nRowsInBlock2, &nRowsInBlock1, (DAAL_INT *)&nFeatures, &negTwo, dataA2, &ldb,
+                                               dataA1, &lda, &zero, mklBuff, &ldc);
+            for (size_t i = 0; i < nRowsInBlock1; ++i)
             {
-                PRAGMA_IVDEP
-                PRAGMA_VECTOR_ALWAYS
-                for (size_t k = 0; k < nVectors2; ++k)
+                algorithmFPType sqrDataA1 = zero;
+                if (a1 != a2)
```

Comment:
just some thoughts:
we could use this statement as template parameter initialization before call `computeInternalVectorVector`:
ex:
```
if (a1 != a2)
  computeInternalVectorVector<true>(...);
else
  computeInternalVectorVector<false>(...);
```

and during call of specific template instantiation of this func we could avoid this condition by compile time optimizations.

Reply:
I redid this `for` a bit. There is no more of this condition.